discovery.go

```diff
@@ -0,0 +1,235 @@
+package pubsub
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// discoverQ handles continuing peer discovery
+	discoverQ chan *discoverReq
+
+	// ongoing tracks ongoing discovery requests
+	ongoing map[string]struct{}
+
+	// done handles completion of a discovery request
+	done chan string
+}
+
+type IsDiscovered func(numPeers int) bool
+
+func CheckNumPeers(num int) IsDiscovered {
+	return func(numPeers int) bool {
+		return numPeers >= num
+	}
+}
+
+// newDiscover creates a new peer discovery mechanism
+func newDiscover() *discover {
+	return &discover{
+		advertising: make(map[string]context.CancelFunc),
+		discoverQ:   make(chan *discoverReq, 32),
+		ongoing:     make(map[string]struct{}),
+		done:        make(chan string),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) {
+	if d.discovery == nil || p == nil {
+		return
+	}
+
+	d.p = p
+	go d.discoverLoop()
+}
+
+func (d *discover) discoverLoop() {
+	for {
+		select {
+		case discover := <-d.discoverQ:
+			topic := discover.topic
+
+			if _, ok := d.ongoing[topic]; ok {
+				discover.done <- struct{}{}
+				continue
+			}
+
+			d.ongoing[topic] = struct{}{}
+
+			go func() {
+				d.handleDiscovery(d.p.ctx, topic, discover.opts)
+				select {
+				case d.done <- topic:
+				case <-d.p.ctx.Done():
+				}
+				discover.done <- struct{}{}
+			}()
+		case topic := <-d.done:
+			delete(d.ongoing, topic)
+		case <-d.p.ctx.Done():
+			return
+		}
+	}
+}
+
+// Advertise advertises this node's interest in a topic to a discovery service. Advertise is not thread-safe.
+func (d *discover) Advertise(topic string) {
+	if d.discovery == nil {
+		return
+	}
+
+	advertisingCtx, cancel := context.WithCancel(d.p.ctx)
```

Comment:
We need to move this below the check below. Technically, not calling cancel could leak a goroutine in some cases.

Reply:
Try running `golangci-lint` against this PR.

Reply:
done